src/standard.ts

```diff
@@ -1,213 +1,71 @@
+import { MultiProof, processProof, processMultiProof } from './core';
+import { MerkleTreeData, MerkleTreeImpl } from './merkletree';
+import { MerkleTreeOptions } from './options';
+import { validateArgument } from './utils/errors';
+
+export type StandardMerkleTreeData<T extends any[]> = MerkleTreeData<T> & {
   format: 'standard-v1';
-  tree: string[];
-  values: {
-    value: T;
-    treeIndex: number;
-  }[];
   leafEncoding: string[];
-}
+};
 
-export class StandardMerkleTree<T extends any[]> {
-  private readonly hashLookup: { [hash: string]: number };
+export function standardLeafHasher<T extends any[]>(types: string[], value: T): HexString {
+  return keccak256(keccak256(defaultAbiCoder.encode(types, value)));
+}
 
-  private constructor(
-    private readonly tree: Bytes[],
-    private readonly values: { value: T; treeIndex: number }[],
-    private readonly leafEncoding: string[],
+export class StandardMerkleTree<T extends any[]> extends MerkleTreeImpl<T> {
+  protected constructor(
+    protected readonly tree: HexString[],
+    protected readonly values: StandardMerkleTreeData<T>['values'],
+    protected readonly leafEncoding: string[],
   ) {
-    this.hashLookup = Object.fromEntries(
-      values.map(({ value }, valueIndex) => [hex(standardLeafHash(value, leafEncoding)), valueIndex]),
-    );
+    super(tree, values, leaf => standardLeafHasher(leafEncoding, leaf));
   }
 
-  static of<T extends any[]>(values: T[], leafEncoding: string[], options: MerkleTreeOptions = {}) {
-    const sortLeaves = options.sortLeaves ?? defaultOptions.sortLeaves;
-
-    const hashedValues = values.map((value, valueIndex) => ({
-      value,
-      valueIndex,
-      hash: standardLeafHash(value, leafEncoding),
-    }));
-
-    if (sortLeaves) {
-      hashedValues.sort((a, b) => compareBytes(a.hash, b.hash));
-    }
-
-    const tree = makeMerkleTree(hashedValues.map(v => v.hash));
-
-    const indexedValues = values.map(value => ({ value, treeIndex: 0 }));
-    for (const [leafIndex, { valueIndex }] of hashedValues.entries()) {
-      indexedValues[valueIndex]!.treeIndex = tree.length - leafIndex - 1;
-    }
-
+  static of<T extends any[]>(
+    values: T[],
+    leafEncoding: string[],
+    options: MerkleTreeOptions = {},
+  ): StandardMerkleTree<T> {
+    const [tree, indexedValues] = MerkleTreeImpl.prepare(values, options, leaf =>
+      standardLeafHasher(leafEncoding, leaf),
+    );
     return new StandardMerkleTree(tree, indexedValues, leafEncoding);
   }
 
   static load<T extends any[]>(data: StandardMerkleTreeData<T>): StandardMerkleTree<T> {
-    if (data.format !== 'standard-v1') {
-      throw new Error(`Unknown format '${data.format}'`);
-    }
-    return new StandardMerkleTree(data.tree.map(hexToBytes), data.values, data.leafEncoding);
+    validateArgument(data.format === 'standard-v1', `Unknown format '${data.format}'`);
+    validateArgument(data.leafEncoding !== undefined, 'Expected leaf encoding');
```

Comment:
These check may appear unecesasry to typescript users, who already get that check from the typing system. But for JS users, this is still needed.